### ISSUE TITLE
MacOS Classic - Procedure definitions after analysis.

### ISSUE THREAD

**Issue by gbody** [uxmal/reko#635]
After running the analyze dataflow on my sample file the the _mapOSerr gets defined as, 
word32 _mapOSerr(word32 d0, word32 d7, word16 wArg06)

Manual analyzing the procedure _mapOSerr, I would have thought the returned value is in D0 and D7 is preserved, so D7 shouldn't be in the parameters.

RTL code generated
----------------------

﻿﻿void _mapOSerr()
{
_mapOSerr_entry:
l00105290:
    a7 = fp
    a5 = a5world
    a7 = a7 - 0x04
    Mem0[a7:word32] = a6
    a6 = a7
    a7 = a7 + 0x00
    a7 = a7 - 0x04
    v5 = d7
    Mem0[a7:word32] = v5
    CVZN = cond(v5)
    d7 = 0x00
    CVZN = cond(d7)
    v8 = Mem0[a6 + 0x0A:word16]
    d0 = DPB(d0, v8, 0)
    CVZN = cond(v8)
    v9 = (word16) d0 - ~0x77
    d0 = DPB(d0, v9, 0)
    CVZNX = cond(v9)
    branch Test(EQ,Z) l001053AA
l001052A4:
    v12 = (word16) d0 - 0x0C
    d0 = DPB(d0, v12, 0)

----------------------------
middle section removed.
----------------------------

l001053AA:
    d7 = 0x0014
    CVZN = cond(d7)
l001053AC:
    d0 = d7
    CVZN = cond(d0)
    d7 = Mem0[a6 + -0x04:word32]
    CVZN = cond(d7)
    a7 = a6
    a6 = Mem0[a7:word32]
    a7 = a7 + 0x04
    return
_mapOSerr_exit:
}

Analyzed procedure
----------------------

word32 _mapOSerr(word32 d0, word32 d7, word16 wArg06)
{
_mapOSerr_entry:
l00105290:
    word32 d7_13 = 0x00
    branch wArg06 == ~0x77 l001053AA
l001052A4:
    branch wArg06 == ~0x6B l0010536E

----------------------------
middle section removed.
----------------------------

l001053AA:
    d7_13 = 0x0014
l001053AC:
    return d7_13
_mapOSerr_exit:
}

**Comment by uxmal** [comment]
Indeed, `d7` is incorrectly marked as live in to the procedure. I will take a peek at this some more this arvo.

**Comment by uxmal** [comment]
That commit removed the ```d0```. Now for the ```d7```....

**Comment by uxmal** [comment]
The gory details: the m68k processor has this annoying "feature" that after a `move.x` instruction, it updates the condition codes in the CPU to reflect the valued that was moved. Reko's m68k rewriter models the m68k statement
```
move.l d7,-(a7)
```
as
```
a7 = a7 - 0x04
v5 = d7
Mem0[a7:word32] = v5
CVZN = cond(v5)
```
Later stages of Reko's analysis can prove the store to `Mem0[a7:word32]` is "dead" and can be removed, but the extra dependency introduced by the `CVZN = cond(v5)` is not eliminated. Thus `v5` is falsely determined to be live, and then the statement `v5 = d7` is also falsely determined to be live, with the end result that `d7` is live.

If I can prove that last CVZN statement is not live, this problem will go away. I will continue looking at this.

**Comment by gbody** [comment]
Looking at the RTL code below, the CVZN = cond(d7) would replace/overwrite the CVZN = cond(v5) and make it a dead end section of code. 
So as long as nothing uses the previous CVZN state then the next CVZN = cond(?) should make the previous CVZN state dead code? Is that right?

a7 = a7 - 0x04
v5 = d7
Mem0[a7:word32] = v5
CVZN = cond(v5)
d7 = 0x00
CVZN = cond(d7)

**Comment by uxmal** [comment]
You're right of course. That definition of `d7` should have killed the use. I'm going to look at this closer tonight to see if I can fix the problem.

**Comment by uxmal** [comment]
Commit 2adb2a2 takes care of the problem in your binary.
